pkg/agent/util/net_windows.go

```diff
@@ -14,7 +14,269 @@
+	if err != nil {
+		return nil, err
+	}
+	return hnsNet, nil
+}
+
+type vSwitchExtensionPolicy struct {
+	ExtensionID string `json:"Id,omitempty"`
+	IsEnabled   bool
+}
+
+type ExtensionsPolicy struct {
+	Extensions []vSwitchExtensionPolicy `json:"Extensions"`
+}
+
+// EnableHNSNetworkExtension enables the specified vSwitchExtension on the target HNS Network. Antrea calls this function
+// to enable OVS Extension on the HNS Network.
+func EnableHNSNetworkExtension(hnsNetID string, vSwitchExtension string) error {
+	extensionPolicy := vSwitchExtensionPolicy{
+		ExtensionID: vSwitchExtension,
+		IsEnabled:   true,
+	}
+	jsonString, _ := json.Marshal(
+		ExtensionsPolicy{
+			Extensions: []vSwitchExtensionPolicy{extensionPolicy},
+		})
+
+	_, err := hcsshim.HNSNetworkRequest("POST", hnsNetID, string(jsonString))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetIPNetDeviceFromIP returns a local IP/mask and associated device from IP.
+func GetIPNetDeviceFromIP(localIP net.IP) (*net.IPNet, *net.Interface, error) {
+	linkList, err := net.Interfaces()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, link := range linkList {
+		addrList, err := link.Addrs()
+		if err != nil {
+			continue
+		}
+		for _, addr := range addrList {
+			if ipNet, ok := addr.(*net.IPNet); ok {
+				if ipNet.IP.Equal(localIP) {
+					return ipNet, &link, nil
+				}
+			}
+		}
+	}
+	return nil, nil, fmt.Errorf("unable to find local IP and device")
+}
+
+func SetLinkUp(name string) (net.HardwareAddr, int, error) {
+	// Set host gateway interface up.
+	if err := EnableHostInterface(name); err != nil {
+		klog.Errorf("Failed to set host link for %s up: %v", name, err)
+		return nil, 0, err
+	}
+
+	iface, err := net.InterfaceByName(name)
+	if err != nil {
+		if opErr, ok := err.(*net.OpError); ok && opErr.Err.Error() == "no such network interface" {
+			return nil, 0, newLinkNoteFoundError(name)
+		}
+		return nil, 0, err
+	}
+	mac := iface.HardwareAddr
+	index := iface.Index
+	return mac, index, nil
+}
+
+func ConfigureLinkAddress(idx int, gwIPNet *net.IPNet) error {
+	iface, _ := net.InterfaceByIndex(idx)
+	gwIP := gwIPNet.IP
+	name := iface.Name
+	if addrs, err := iface.Addrs(); err != nil {
+		klog.Errorf("Failed to query IPv4 address list for interface %s: %v", name, err)
+		return err
+	} else if addrs != nil {
+		for _, addr := range addrs {
+			// Check with IPv4 address.
+			if ipNet, ok := addr.(*net.IPNet); ok {
+				if ipNet.IP.To4() != nil && ipNet.IP.Equal(gwIPNet.IP) {
+					return nil
+				}
+			}
+		}
+	}
+
+	klog.V(2).Infof("Adding address %v to gateway interface %s", gwIP, name)
+	if err := ConfigureInterfaceAddress(iface.Name, gwIPNet); err != nil {
+		klog.Errorf("Failed to set gateway interface %s with address %v: %v", iface, gwIP, err)
+		return err
+	}
+	return nil
+}
+
+// PrepareHostNetworking creates HNS Network for containers.
+func PrepareHostNetworking(subnetCIDR *net.IPNet, nodeIPNet *net.IPNet) error {
```

Comment:
I am not sure this one and enableHNSOnOVS() should be put inside package util. They sound like some application logics but not utilities?
Should we still have the agent_windows.go? @tnqn

Reply:
According to the offline disscussion we had last week, I remove the call of function `PrepareHostNetworking` from agent.go, and plan to move it to Antrea-Service, that's why I move the function to utils(If it is still not proper, I think a choice is to move it to Antrea-Service). Then the only function differentiates OS in agent should be SNAT. Quan has moved iptables-SNAT to routeClient initialization. Windows is using OpenFlow to provide this feature, and I didn't want to add OFClient as a property of routeClient. So agent_window might be added back in "SNAT" PR.

Reply:
Sorry, you mean you want to move PrepareHostNetworking() to Service later, or keep it in Agent? If we keep it Agent, I feel it should not be in util, and agent_windows.go is better.

Reply:
It is planned to move to Service.

Reply:
Do we have other similar funcs in service code?
In dev env, if we do not have a service, how could we prepare the network before Agent runs?

Reply:
The service is also a binary file written with go, and provided in cmd/. In dev env, we could run the service binary one time, but not register as a Windows Service. Then the binary could create HNS Network, and run Agent once. We could interrupt the service process, and then run/debug Agent code.

Reply:
Got it. I now feel from convenience perspective, seems the logic in agent is more convenient. So, I wonder do we have similar cases in service now (I remember no).
If we anyway will have agent_windows.go for other funcs like NAT, we might put this one there too?

Reply:
Currently, we only have three functions planned to be in Service: 1) Start/Monitor Agent, 2) Prepare environment including HNS Network; 3) Monitoring ovs-vswitchd service, and recover host network when OVS is down.

No other funcs like `PrepareHostNetworking` are in Service.

A bad effect for placing PrepareHostNetworking in Agent is, it will make Agent behavior on Windows different from Linux.

Reply:
I assume 2) and 3) (recover host network when OVS is down) are not really relevant right, or 3) is to recover what are changed in 2)?

Reply:
In my opinion, the latter is a correct relation between 2) and 3), as OVS takes over Windows host network is because of 2), and when OVS is down, the host network should be controlled by Windows system.

Reply:
Ok. Then it makes to move them to service.